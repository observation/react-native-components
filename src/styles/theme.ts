import Color from 'color'

const font = {
  size: {
    extraSmall: 10,
    small: 12,
    medium: 14,
    large: 16,
    extraLarge: 18,
    huge: 31,
  },
  lineHeight: {
    extraSmall: 16,
    small: 17,
    medium: 20,
    large: 26,
    extraLarge: 28,
    huge: 37,
  },
}

const icon = {
  size: {
    extraSmall: 10,
    small: 12,
    medium: 14,
    large: 16,
    extraLarge: 18,
    extraExtraLarge: 24,
    huge: 48,
  },
}

const black = '#212121'
const greyBase = '#939393'

const color = {
  primary: '#0066B1',
  primaryLighter: '#66A3D0',
  primaryLightest: '#E6F0F7',
  secondaryLight: '#709FFD',
  secondaryLightestExtra: '#F0F5FF',
  black,
  greyDark: '#666666',
  greyBase,
  placeholder: greyBase,
  greySemi: '#E6E6E6',
  greyLight: '#F8F9FA',
  white: '#FFFFFF',
  white70: '#FFFFFFB4',
  white77: '#FFFFFFC4',
  white80: '#FFFFFFCD',
  success: '#84B72C',
  successLight: '#9DC556',
  successLighter: '#B5D480',
  successLightest: '#CEE2AB',
  successLightestExtra: '#F7FBEF',
  successDark: '#6A9223',
  warning: '#F4C013',
  error: '#EA554B',
  errorLightest: '#F7BBB7',
  modalBackground: Color(black).alpha(0.6).lighten(0.4).string(),
  darkPanelBackground: Color(black).alpha(0.5).string(),
}

const margin = {
  eighth: 2,
  quarter: 4,
  half: 8,
  common: 16,
  large: 24,
  double: 32,
  huge: 48,
}

const rounded = {
  borderRadius: 4,
  overflow: 'hidden' as 'hidden',
}

const input = {
  ...rounded,
  flex: 1,
  minHeight: 40,
  borderWidth: 2,
  paddingLeft: margin.half,
  paddingRight: margin.double,
}

const bottomGradientColors = ['#30303000', '#30303059']
const topGradientColors = bottomGradientColors.slice().reverse()

const getBorderColor = ({ isFocused = false, hasErrors = false }) =>
  hasErrors ? color.error : isFocused ? color.primaryLighter : color.greySemi

export default {
  font,
  icon,
  color,
  bottomGradientColors,
  topGradientColors,
  margin,
  input,
  shadow: {
    android: {
      elevation: 6,
      shadowColor: color.greyBase,
    },
    ios: {
      shadowOffset: { width: 0, height: 4 },
      shadowColor: color.greySemi,
      shadowOpacity: 0.8,
      shadowRadius: 4,
    },
  },
  shadowSmall: {
    android: {
      elevation: 4,
      shadowColor: color.greyBase,
    },
    ios: {
      shadowOffset: { width: 0, height: 2 },
      shadowColor: color.greySemi,
      shadowOpacity: 0.8,
      shadowRadius: 2,
    },
  },
  rounded,
  roundedLarge: {
    borderRadius: 8,
    overflow: 'hidden' as 'hidden',
  },
  roundedHuge: {
    borderRadius: 10,
    overflow: 'hidden' as 'hidden',
  },
  absolute: {
    position: 'absolute' as 'absolute',
    top: 0,
    bottom: 0,
    left: 0,
    right: 0,
  },
  absoluteLeft: {
    position: 'absolute' as 'absolute',
    top: 0,
    bottom: 0,
    left: 0,
  },
  absoluteRight: {
    position: 'absolute' as 'absolute',
    top: 0,
    bottom: 0,
    right: 0,
  },
  absoluteBottom: {
    position: 'absolute' as 'absolute',
    bottom: 0,
    left: 0,
    right: 0,
  },
  absoluteTop: {
    position: 'absolute' as 'absolute',
    top: 0,
    left: 0,
    right: 0,
  },
  getBorderColor,
}
