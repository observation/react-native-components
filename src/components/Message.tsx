import React from 'react'
import { StyleSheet, View, Text, ViewStyle, StyleProp } from 'react-native'

import LargeButton, { LargeButtonProps } from '../components/LargeButton'
import textStyle from '../styles/text'
import theme from '../styles/theme'

type Props = {
  title?: string
  text: string
  buttons?: LargeButtonProps[]
  style?: StyleProp<ViewStyle>
}

const Message = ({ title, text, buttons, style }: Props) => (
  <View style={[styles.message, style]}>
    {title && (
      <View style={{ marginBottom: theme.margin.half }}>
        <Text style={textStyle.lead}>{title}</Text>
      </View>
    )}
    <View>
      <Text style={textStyle.body}>{text}</Text>
    </View>
    {buttons && buttons.length > 0 && (
      <View
        style={{
          marginTop: theme.margin.common,
          flexDirection: 'row',
          alignItems: 'flex-start',
        }}
      >
        {buttons.map((button, i) => (
          <LargeButton
            disabled={button!.disabled || false}
            secondary={button!.secondary || false}
            title={button!.title}
            key={i}
            style={{ margin: 0, marginRight: theme.margin.half }}
            onPress={button!.onPress!}
          />
        ))}
      </View>
    )}
  </View>
)

export default Message

const styles = StyleSheet.create({
  message: {
    flexDirection: 'column',
    marginBottom: 0,
  },
})
