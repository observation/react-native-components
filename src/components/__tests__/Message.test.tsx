import React from 'react'

import { render } from '@testing-library/react-native'

import theme from '../../styles/theme'
import Message from '../Message'

jest.mock('../../components/LargeButton', () => 'mock-large-button')

describe('Message', () => {
  describe('Rendering', () => {
    test('Without button', () => {
      // GIVEN
      const { toJSON } = render(<Message title="Title" text="Text" />)

      // THEN
      expect(toJSON()).toMatchSnapshot()
    })

    test('With button', () => {
      // GIVEN
      const { toJSON } = render(
        <Message title="Title" text="Text" buttons={[{ onPress: () => {}, title: 'button title' }]} />,
      )

      // THEN
      expect(toJSON()).toMatchSnapshot()
    })

    test('Without title', () => {
      // GIVEN
      const { toJSON } = render(<Message text="Text" buttons={[{ onPress: () => {}, title: 'button title' }]} />)

      // THEN
      expect(toJSON()).toMatchSnapshot()
    })

    test('With style', () => {
      // GIVEN
      const { toJSON } = render(<Message title="Title" text="Text" style={{ margin: theme.margin.common }} />)

      // THEN
      expect(toJSON()).toMatchSnapshot()
    })
  })
})
