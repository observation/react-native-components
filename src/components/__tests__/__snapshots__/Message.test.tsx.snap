// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Message Rendering With button 1`] = `
<View
  style={
    [
      {
        "flexDirection": "column",
        "marginBottom": 0,
      },
      undefined,
    ]
  }
>
  <View
    style={
      {
        "marginBottom": 8,
      }
    }
  >
    <Text
      style={
        {
          "color": "#212121",
          "fontFamily": "Ubuntu",
          "fontSize": 14,
          "fontStyle": "normal",
          "fontWeight": "bold",
          "lineHeight": 20,
        }
      }
    >
      Title
    </Text>
  </View>
  <View>
    <Text
      style={
        {
          "color": "#666666",
          "fontFamily": "Ubuntu",
          "fontSize": 14,
          "fontStyle": "normal",
          "fontWeight": "normal",
          "lineHeight": 20,
        }
      }
    >
      Text
    </Text>
  </View>
  <View
    style={
      {
        "alignItems": "flex-start",
        "flexDirection": "row",
        "marginTop": 16,
      }
    }
  >
    <mock-large-button
      disabled={false}
      onPress={[Function]}
      secondary={false}
      style={
        {
          "margin": 0,
          "marginRight": 8,
        }
      }
      title="button title"
    />
  </View>
</View>
`;

exports[`Message Rendering With style 1`] = `
<View
  style={
    [
      {
        "flexDirection": "column",
        "marginBottom": 0,
      },
      {
        "margin": 16,
      },
    ]
  }
>
  <View
    style={
      {
        "marginBottom": 8,
      }
    }
  >
    <Text
      style={
        {
          "color": "#212121",
          "fontFamily": "Ubuntu",
          "fontSize": 14,
          "fontStyle": "normal",
          "fontWeight": "bold",
          "lineHeight": 20,
        }
      }
    >
      Title
    </Text>
  </View>
  <View>
    <Text
      style={
        {
          "color": "#666666",
          "fontFamily": "Ubuntu",
          "fontSize": 14,
          "fontStyle": "normal",
          "fontWeight": "normal",
          "lineHeight": 20,
        }
      }
    >
      Text
    </Text>
  </View>
</View>
`;

exports[`Message Rendering Without button 1`] = `
<View
  style={
    [
      {
        "flexDirection": "column",
        "marginBottom": 0,
      },
      undefined,
    ]
  }
>
  <View
    style={
      {
        "marginBottom": 8,
      }
    }
  >
    <Text
      style={
        {
          "color": "#212121",
          "fontFamily": "Ubuntu",
          "fontSize": 14,
          "fontStyle": "normal",
          "fontWeight": "bold",
          "lineHeight": 20,
        }
      }
    >
      Title
    </Text>
  </View>
  <View>
    <Text
      style={
        {
          "color": "#666666",
          "fontFamily": "Ubuntu",
          "fontSize": 14,
          "fontStyle": "normal",
          "fontWeight": "normal",
          "lineHeight": 20,
        }
      }
    >
      Text
    </Text>
  </View>
</View>
`;

exports[`Message Rendering Without title 1`] = `
<View
  style={
    [
      {
        "flexDirection": "column",
        "marginBottom": 0,
      },
      undefined,
    ]
  }
>
  <View>
    <Text
      style={
        {
          "color": "#666666",
          "fontFamily": "Ubuntu",
          "fontSize": 14,
          "fontStyle": "normal",
          "fontWeight": "normal",
          "lineHeight": 20,
        }
      }
    >
      Text
    </Text>
  </View>
  <View
    style={
      {
        "alignItems": "flex-start",
        "flexDirection": "row",
        "marginTop": 16,
      }
    }
  >
    <mock-large-button
      disabled={false}
      onPress={[Function]}
      secondary={false}
      style={
        {
          "margin": 0,
          "marginRight": 8,
        }
      }
      title="button title"
    />
  </View>
</View>
`;
